[pytorch] add irregular c10 op registration/invocation cases to test project

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30558 [pytorch] add irregular c10 op registration/invocation cases to test project**

Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18811098](https://our.internmc.facebook.com/intern/diff/D18811098)

ljk53 committed:
[pytorch] add irregular c10 op registration/invocation cases to test project

Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add irregular c10 op registration/invocation cases to test project"


Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18742750](https://our.internmc.facebook.com/intern/diff/D18742750)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add irregular c10 op registration/invocation cases to test project"


Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18742750](https://our.internmc.facebook.com/intern/diff/D18742750)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add irregular c10 op registration/invocation cases to test project"


Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18742750](https://our.internmc.facebook.com/intern/diff/D18742750)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add irregular c10 op registration/invocation cases to test project"


Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18742750](https://our.internmc.facebook.com/intern/diff/D18742750)

[ghstack-poisoned]

ezyang reviewed (approved):
OK.

@smessmer is there any way we can make the calling sites here more regular? It doesn't seem to me like there should be a distinction here.

ljk53 commented:
> OK.
> 
> @smessmer is there any way we can make the calling sites here more regular? It doesn't seem to me like there should be a distinction here.

To be clear I introduced 3 irregular cases in this PR:

1) How it calls other ops. The only case that doesn't directly call "c10::Dispatcher::findSchema" is the "callOp" helper we created in ATen/native/c10_utils.h. We can refactor the code to get rid of this exception case but similar cases might be introduced in future unless we set the convention/lint rule somewhere that "findSchema" needs to be called directly or always call other ops via generated native interface.

2) How it registers ops without using "c10::RegisterOperators::Options". The codegen created registration uses ".op(torch::RegisterOperators::options().schema("aten::AA(Tensor self) -> Tensor").kernel<decltype(AA_op), &AA_op>(TensorTypeId::CPUTensorId)" while some manually written registration uses ".op("aten::AA", &AA_op)" directly. It broke my original assumption that registrations are interlaced by Options::schema call, so I changed the code analyzer to do better flow analysis.

3) How it passes parameters to kernel as template argument. I first noticed it in quantization code where we passed in "ReLUFused" boolean: "QAdd</*ReLUFused=*/true>>". The original code analyzer could handle it correctly. I introduced a hypothetical case where called op name is passed in as template argument just to test the robustness of the new analyzer - but we don't have such use cases in aten (yet).

2 & 3 are more on registration / kernel side and only 1 is related to calling site.

ljk53 committed:
Update on "[pytorch] add irregular c10 op registration/invocation cases to test project"


Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18742750](https://our.internmc.facebook.com/intern/diff/D18742750)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add irregular c10 op registration/invocation cases to test project"


Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18742750](https://our.internmc.facebook.com/intern/diff/D18742750)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add irregular c10 op registration/invocation cases to test project"


Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18742750](https://our.internmc.facebook.com/intern/diff/D18742750)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add irregular c10 op registration/invocation cases to test project"


Summary:
Most c10 op registration/invocation cases are generated by aten codegen
following some fixed pattern, but a handful of them were written
manually, mainly for quantized ops. Added these "irregular" cases to the
test project to verify static code analyzer can handle them as well.

Test:
- build and run the test project;

Differential Revision: [D18742750](https://our.internmc.facebook.com/intern/diff/D18742750)

[ghstack-poisoned]
